### ISSUE TITLE
"Warning: 'Example already has an EOS token appended' and Increased Memory Usage with Mistral Nemo Model"

### ISSUE THREAD

**Issue by Satyam7166-tech** [ml-explore/mlx-examples#944]
I get the warning even after validating the whole dataset through [mistral-finetune](https://github.com/mistralai/mistral-finetune)

The model is still training so will share the training log after it completes, to know if this warning is effecting my model. Though I am not sure why I get the warning in the first place.

The issues persists even if I  use the latest python package of mlx_lm or build it from source.

My sample data:

`{"messages": [{"role": "user", "content": "<|begin_of_text|><|start_header_id|>user<|end_header_id|>\\n\\nFollow task: Lorem Ipsum."}, {"role": "assistant", "content": "assistant<|end_header_id|>\\n\\nLorem Ipsum.<|end_of_text|>"}]}`

Model used: `mistralai/Mistral-Nemo-Instruct-2407`

Command: 

(Edited) Trace:

> python -m mlx_lm.lora \
>   --train \
>   --model mistralai/Mistral-Nemo-Instruct-2407 \
>   --data /Users/macstudiosct/projects/finetune_mlx/data  \
>   --batch-size 3 \
>   --lora-layers -1 \
>   --iters 8000 --max-seq-length 8000 --adapter-path finetuning_nemo_refresh/adapters
> Loading pretrained model
> Fetching 16 files: 100%|███████████████████████████████████████████████████████████████████████████| 16/16 [00:00<00:00, 20648.88it/s]
> Loading datasets
> Training
> Trainable parameters: 0.040% (4.915M/12247.782M)
> Starting training..., iters: 8000
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> [WARNING] Example already has an EOS token appended
> Iter 1: Val loss 2.604, Val took 113.465s
> [WARNING] Example already has an EOS token appended

**Comment by Satyam7166-tech** [comment]
Also my memory consumption keeps increasing after every iteration:

Added txt file for readability:

[trace.txt](https://github.com/user-attachments/files/16648004/trace.txt)

**Comment by awni** [comment]
The MLX LM data preprocessor will add an EOS token to the text before training on it. In your case it is finding an eos token there already so it won't add the EOS token. It gives that warning instead and does nothing. You can safely ignore the warning in that case.. or you can preprocess your data to not include an EOS token to avoid the warning. (Maybe we should not be so verbose with that warning, I'll think on that..)

**Comment by Satyam7166-tech** [comment]
> The MLX LM data preprocessor will add an EOS token to the text before training on it. In your case it is finding an eos token there already so it won't add the EOS token. It gives that warning instead and does nothing. You can safely ignore the warning in that case.. or you can preprocess your data to not include an EOS token to avoid the warning. (Maybe we should not be so verbose with that warning, I'll think on that..)

I see, thanks.

Though can you have a look at the txt file. Am I imagining things or is the memory "leak" an issue here?

**Comment by awni** [comment]
So the peak memory use is determined by the longest example seen so far. In your case what's likely happening is it's going up every now and then as you encounter a longer sequence than any sequence you've seen so far during training.

You could verify that there is no leak by setting `--max-seq-length 512` and observing that the memory should be constant after the first sequence you encounter which has length greater or equal to 512.

The length `8000` is very high.. so you want to be careful with that as it can consume a lot of memory due to the quadratic attention.

**Comment by Satyam7166-tech** [comment]
> So the peak memory use is determined by the longest example seen so far. In your case what's likely happening is it's going up every now and then as you encounter a longer sequence than any sequence you've seen so far during training.
> 
> 
> 
> You could verify that there is no leak by setting `--max-seq-length 512` and observing that the memory should be constant after the first sequence you encounter which has length greater or equal to 512.
> 
> 
> 
> The length `8000` is very high.. so you want to be careful with that as it can consume a lot of memory due to the quadratic attention.

Oh I see. Thank you so much @awni. I learn a lot by working with this library. It is literally life changing :)

I'll close the issue as it has been resolved. Though if possible, can you explain what sort of quadratic equation takes place here, if you have the time.

**Comment by awni** [comment]
> Though if possible, can you explain what sort of quadratic equation takes place here, if you have the time.

To compute the attention scores you do  this:

```
# queries, keys have shape [sequence length, dim]
scores = queries @ keys.T
```

The scores have shape `[sequence length, sequence length]` which scales with the square of the sequence length.
